### PR TITLE
[zlaski/memset-model] QL model for `memset` and friends

### PR DESCRIPTION
**Pull request by zlaski-semmle**

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.DataFlow
+import semmle.code.cpp.models.interfaces.Taint
+
+/**
+ * The standard function `memset` and its assorted variants
+ */
+class MemsetFunction extends ArrayFunction, DataFlowFunction, TaintFunction {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: You can remove `TaintFunction` from the list of superclasses and remove `Taint` from the list of imports.

Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: I think `AliasFunction` should be extended as well in order to tell the escape analysis that `memset` doesn't retain a pointer for later. See https://github.com/Semmle/ql/blob/0db648beaeba490ee29d00104d254c27c83f65ba/cpp/ql/src/semmle/code/cpp/models/interfaces/Alias.qll#L43-L48

File: cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  MemsetFunction() {
+    hasGlobalName("memset") or
+    hasGlobalName("bzero") or
+    hasGlobalName("__builtin_memset") or
+    hasQualifiedName("std", "memset")
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: `wmemset`?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: Or add taint flow from the character and number parameters to the buffer???  An attacker with control of them has some (very limited) control of the data that ends up in the buffer and could plausibly use this for harm.

**Comment by geoffw0** [comment]
LGTM.  Most of our suggestions are actually requests for improvements, not fixes, so feel free to defer them as future work.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: I argued against that in https://github.com/Semmle/ql/pull/1933#discussion_r327962404.

@zlaski-semmle and I just had a discussion about what constitutes taint. Here is the definition I wrote a long time ago: https://github.com/Semmle/ql/blob/e1594a4b0b9ff0cd3fe70e873ad38a98c0a6b41d/cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking.qll#L5-L8. For other languages, like Java, it's completely different, and they haven't attempted to write down a definition. I think the only "definition" that fits all languages is: _taint should flow where it's beneficial for some taint-tracking queries and harmful for none_.

Comment (zlaski-semmle) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: Done.  Let me know if I've over-specified things :-)

File: cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    hasGlobalName("__builtin_memset_chk") or
+    hasQualifiedName("std", "memset") or
+    hasQualifiedName("std", "wmemset")
+  }
+
+  override predicate hasArrayOutput(int bufParam) { bufParam = 0 }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isInParameter(0) and
+    output.isOutReturnValue()
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Memset.qll]: I've just merged #1938, so these names are now deprecated.

**Comment by jbj** [comment]
This PR LGTM now. It needs
- Geoffrey's review,
- qlformat, and
- to be taken out of Draft state.

**Comment by zlaski-semmle** [comment]
I believe this PR is now mergeable.  I'd like to get it merged so that I may continue work on https://github.com/Semmle/ql/pull/1933.

**Comment by jbj** [comment]
It looks like @geoffw0 has no objections, so I'll merge this.

**Pull request merged**